Transformers: Add smoothing transformer

vesalaakso-oura opened the pull request:
**What is this feature?**

I built a smoothing transformer that cleans up noisy time series data. It uses the ASAP algorithm to pick which data points to keep. This works similarly to Datadog's `autosmooth()` function.

**Why do we need this feature?**

This was actually requested back in 2019 in [issue #15022](https://github.com/grafana/grafana/issues/15022). I was migrating dashboards from Datadog to Grafana and we noticed that Datadog’s autosmoothing feature, which made graphs easier to read was not available in Grafana, so we had to find a workaround.. 

**Who is this feature for?**

Users coming from Datadog who miss the autosmooth functionality

**Which issue(s) does this PR fix?**:

Fixes #15022

**Special notes for your reviewer:**

I've included a gif demo of the feature in action 😎 

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle. (N/A - standard transformer)
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc. **NOTE, I believe only maintainers can update the CMS.

**What I built:**
- Smoothing transformer using the ASAP downsampling algorithm (same one mentioned in the original issue)
- Always keeps the first and last data points so you get the full time range (otherwise graph might look incomplete)
- Simple resolution setting (1-1000 points) to control how aggressive the smoothing is
- Interpolate the smoothed curve back to the original time points making sure the output matches the original timestamps exactly.
- Test cases covering all the edge cases I could think of
- Docs with examples

**Demo gif:**
![pr](https://github.com/user-attachments/assets/b25c3d27-cc46-450b-b902-becb41a4345e)

vesalaakso-oura committed:
Transformers: Add smoothing transformer

Added a smoothing transformer to help clean up noisy time series data.
It uses the ASAP algorithm to pick the most important data points while
keeping the overall shape and trends intact.

The transformer always keeps the first and last points so you get the
complete time range. I also added a test for it.

kylebrandt commented:
Another possible option using SQL Expressions, keeps spike while smoothing. Might not be as robust as the autosmoother example, and takes a couple more params:

<img width="930" height="1239" alt="image" src="https://github.com/user-attachments/assets/fa425556-ad7c-499c-b684-c1dd700cac59" />

<img width="593" height="1128" alt="image" src="https://github.com/user-attachments/assets/7f65e9d1-e75d-409a-a7a1-f002b88237e7" />

```
WITH w AS (
  SELECT
    time,
    a,
    __value__  AS val,

    /* rolling center & spread over the last 20 rows */
    AVG(__value__) OVER (
      PARTITION BY a
      ORDER BY time
      ROWS BETWEEN 20 PRECEDING AND CURRENT ROW
    ) AS mu,

    STDDEV_POP(__value__) OVER (
      PARTITION BY a
      ORDER BY time
      ROWS BETWEEN 20 PRECEDING AND CURRENT ROW
    ) AS sigma,

    /* neighbors for "extra-picky" local-extremum check */
    LAG(__value__, 1)  OVER (PARTITION BY a ORDER BY time) AS prev_val,
    LEAD(__value__, 1) OVER (PARTITION BY a ORDER BY time) AS next_val
  FROM A
)
SELECT
  time,
  a,
  val AS Value,
  CASE
    WHEN sigma IS NULL OR sigma = 0 THEN mu
    /* keep spikes/dips when z >= 2.5 and it's a local extremum (or at edges) */
    WHEN ABS(val - mu) >= 2 * sigma
         AND (
              prev_val IS NULL OR next_val IS NULL
              OR (val > prev_val AND val > next_val)
              OR (val < prev_val AND val < next_val)
         )
      THEN val
    ELSE mu
  END AS smoothed
FROM w
ORDER BY a, time;
```

vesalaakso-oura commented:
Hi @kylebrandt , thanks a lot for taking the time to write up the SQL approach and sharing the example. I really appreciate the effort you put into it. 🙇 

My main concern is scalability in larger orgs. If every team needs to ship or maintain their own SQL query, it quickly becomes a burden. It also makes it harder to convince teams migrating from datadog, where autosmoothing is just available out of the box. Having a built-in ASAP transformation would not only help internally but could also tip other organisations out there to consider moving away from datadog to the grafana stack.

ASAP itself is well researched. The original [paper](https://kexinrong.github.io/papers/asap-vldb17.pdf) shows that ASAP can spot meaningful trends and outliers while suppressing noise, so having it as a built-in transformation would be an easier sell.

vesalaakso-oura commented:
Hello @gelicia, is this a feature we can proceed with, or do you have any concerns or alternative ways to implement it?

vesalaakso-oura commented:
@gelicia Hello, I hope you are doing well. Any news so far?

samjewell commented:
Hey @vesalaakso-oura, thanks for this.

I wanted to ask whether you had considered adding this to Prometheus instead? If you have, can you say why it makes sense to add this to Grafana Transforms, and not there?

To me this feels like it could fit more seamlessly into Prometheus for a few reasons:
- It feels metrics specific. I'm aware Grafana is "big tent", and so putting it into Transforms would allow this to be used with other metrics data sources. Yet it still doesn't feel like a generic transform to me, but specific to the metrics domain. If we add this to Prom then all Prom users can benefit, not just Grafana users.
- Grafana Transforms are applied in the front-end only, and cannot be used in either **alerts** or **recording rules** (whether evaluated by Alertmanager or Grafana). 
- As a user, calling a function with a name such as this, I feel like I want to reach for this in PromQL, and not in Grafana Transforms.

I also wondered if you could share more of your own context? You said:
> makes it harder to convince teams migrating from datadog

Is this your context- are you trying to remove blockers in order to migrate teams over from Datadog? If this is the context, I feel like an implementation in PromQL could be just as good, or better in fact than adding a Transform, as a DDog migration would typically move metrics into our hosted-prometheus offering. I feel like the UX of a function in PromQL is simpler than one where you have to switch tabs, add a transform, and then find the one you want.

What do you think? Thanks 🙏

Develer reviewed (commented):
I don’t expect to see dots instead of a line after applying ASAP. It looks like the current implementation just adds many null points, which is why it appears as dots - but I need to look into it more closely.

<img width="1183" height="1220" alt="Image" src="https://github.com/user-attachments/assets/407167fa-e073-41db-9431-51dd881a7d29" />
<img width="806" height="306" alt="Image" src="https://github.com/user-attachments/assets/69d0d728-c692-4243-be90-77ee5f400e96" />

Develer commented on public/app/features/transformers/smoothing/smoothing.ts:
Currently it's defined inside the transformer closure on every call. Maybe we could lift this function up?

Develer commented on public/app/features/transformers/smoothing/smoothing.test.ts:
Maybe it's a typo? This comment looks misleading since resolution 2 is already above the minimum (1).

Develer reviewed (approved):
@vesalaakso-oura
This PR looks great - thank you for your passion and effort on this smoothing transformation! 🎉
I've left a few minor suggestions/requested changes, but overall this is an approve from my side.

**Just a heads up**: we still have a few open questions to discuss with another team regarding this PR. @alexjonspencer1 will handle the next steps and coordinate from here.
Thanks again for this contribution! 🙏

alexjonspencer1 commented:
## Proposed transformation icons

### Dark:
![smoothing-dark](https://github.com/user-attachments/assets/2b7d521c-2279-4c96-910f-4c1bf3a325c8)

### Light: 
![smoothing-light](https://github.com/user-attachments/assets/2af3895f-18bc-4b6a-a204-45b271464111)

vesalaakso-oura committed:
Change category

Change category from Reformat to CalculateNewFields

vesalaakso-oura committed:
Remove first/last point preservation

vesalaakso-oura committed:
Fix operator recreation

vesalaakso-oura committed:
Simplify ASAP code

Include performance optimization as well

vesalaakso-oura committed:
Refactor interpolateFromSmoothedCurve

Break function into smaller focused functions and lift functions to the
top level

vesalaakso-oura committed:
Add isApplicable Check

Make sure the transformer is applicable for timeseries data

vesalaakso-oura committed:
Add tests for isApplicable check

vesalaakso-oura committed:
UI/UX improvements: Display effective resolution when limited by data points

Show "Effective: X" indicator when resolution is capped by the 2x data
points multiplier. Includes tooltip explaining the limit.

Memoizes calculation to prevent unnecessary recalculation on re-renders.

Example: With 72 data points and resolution set to 150, displays
"Effective: 144" since the limit is 72 x 2 = 144.

Plus added tests

vesalaakso-oura committed:
Improve discoverability by adding tags

vesalaakso-oura committed:
Preserve Original Data

Let's preserve original data as well, makes the UX so much better.
Changed from appending (smoothed) to frame names to use Smoothed frame name. This should match the pattern used by other transformers (e.g,. regression)
Updated tests accordingly
Updated tooltip note

vesalaakso-oura committed:
Add asap tests

Basic functionality:
* returns valid DataPoint objects
* Maintain x-axis ordering

Edge cases:
* Empty array
* single data point
* filter NaN values
* all NaN values
* sort unsorted data
* negative values

vesalaakso-oura committed:
Update dark and light images

alexjonspencer1 commented on public/app/features/transformers/smoothing/smoothing.ts:
**Nit (non-blocking):** Might be a good idea to clear state cache. We've run into bugs with this before. 

```suggestion
smoothedFields.push({
  ...field,
  values: smoothedValues,
  state: undefined
});
```

alexjonspencer1 replied:
Thanks @vesalaakso-oura!! I think the only remaining feedback is the comment above addressing 1) feature toggle, 2) running a translation script, and 3) resolving conflicts.

vesalaakso-oura committed:
Clear state cache

vesalaakso-oura committed:
Add feature toggle

vesalaakso-oura committed:
Conditionally add new transformation to the registry

alexjonspencer1 commented:
Hey @vesalaakso-oura. Getting super close on this, thanks for your continued patience! The branch needs to be rebased on the latest `main` before we can merge. 

The current PR includes massive diffs in generated files (1,990+ lines in `toggles_gen.json`) because the feature toggles were generated against an older version of main.

Can you please:
1. Rebase your branch on latest `main`
2. Re-run `make gen-feature-toggles`

This will reduce the diff to only your changes. Thanks!

vesalaakso-oura committed:
chore: update and regenerate feature toggles

alexjonspencer1 commented on public/app/features/transformers/standardTransformers.ts:
```diff
@@ -64,8 +65,9 @@ export const getStandardTransformers = (): TransformerRegistryItem[] => {
-    getFormatStringTransformerRegistryItem(),
-    getGroupToNestedTableTransformRegistryItem(),
+    ...(config.featureToggles.formatString ? [getFormatStringTransformerRegistryItem()] : []),
+    ...(config.featureToggles.groupToNestedTableTransformation ? [getGroupToNestedTableTransformRegistryItem()] : []),
```
Looks like this was accidentally made conditional.

```suggestion
    getGroupToNestedTableTransformRegistryItem(),
```

alexjonspencer1 commented on yarn.lock:
```diff
@@ -1466,20 +1466,13 @@ __metadata:
   linkType: hard
 
-"@babel/runtime@npm:7.28.2":
+"@babel/runtime@npm:7.28.2, @babel/runtime@npm:^7.0.0, @babel/runtime@npm:^7.1.2, @babel/runtime@npm:^7.10.1, @babel/runtime@npm:^7.11.1, @babel/runtime@npm:^7.11.2, @babel/runtime@npm:^7.12.0, @babel/runtime@npm:^7.12.1, @babel/runtime@npm:^7.12.13, @babel/runtime@npm:^7.12.5, @babel/runtime@npm:^7.13.10, @babel/runtime@npm:^7.16.3, @babel/runtime@npm:^7.17.8, @babel/runtime@npm:^7.18.0, @babel/ ...
```
This diff in the `package.lock` seems incorrect. This diff should be much smaller given the only addition was the `downsample` dependency.

vesalaakso-oura committed:
chore: update yarn.lock

vesalaakso-oura committed:
chore: fix transformers and imports

Develer reviewed (approved):
Hi @vesalaakso-oura,
@alexjonspencer1 is on PTO, so I'll push this PR to the finish line with you.

> I think the only remaining feedback is the comment above addressing 1) feature toggle, 2) running a translation script, and 3) resolving conflicts.

- [x] feature toggle
- [x] translations
- [x] resolving conflicts

I double-checked transformation functionality locally, and things look good to me 🚀🚀🚀
Thanks again for your contribution 🙌🏻
Feel free to merge it 🚢

vesalaakso-oura commented:
> Hi @vesalaakso-oura, @alexjonspencer1 is on PTO, so I'll push this PR to the finish line with you.
> 
> > I think the only remaining feedback is the comment above addressing 1) feature toggle, 2) running a translation script, and 3) resolving conflicts.
> 
> * [x]  feature toggle
> * [x]  translations
> * [x]  resolving conflicts
> 
> I double-checked transformation functionality locally, and things look good to me 🚀🚀🚀 Thanks again for your contribution 🙌🏻 Feel free to merge it 🚢

Thanks for the approval @Develer ! I'd love to merge but it looks like I don't have write permissions to this branch (merging is blocked for me).

I believe I've addressed the feedback regarding the feature toggle, translations, and conflicts. Could you please trigger the merge on your end? 🚢
